custom_components/sems/sensor.py

```diff
@@ -5,38 +5,115 @@
-    CoordinatorEntity,
-    DataUpdateCoordinator,
-    UpdateFailed,
+from homeassistant.components.sensor import (
+    SensorEntity,
+    SensorDeviceClass,
+    SensorStateClass,
 )
-from homeassistant.components.sensor import STATE_CLASS_TOTAL_INCREASING, SensorEntity
+from homeassistant.config_entries import ConfigEntry
 from homeassistant.const import (
-    DEVICE_CLASS_POWER,
     POWER_WATT,
     CONF_SCAN_INTERVAL,
-    DEVICE_CLASS_ENERGY,
     ENERGY_KILO_WATT_HOUR,
+    TEMP_CELSIUS,
+    ELECTRIC_POTENTIAL_VOLT,
+    ELECTRIC_CURRENT_AMPERE,
+    FREQUENCY_HERTZ,
+    PERCENTAGE,
+    POWER_KILO_WATT,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+    UpdateFailed,
+)
+
+from .const import (
+    DOMAIN,
+    CONF_STATION_ID,
+    DEFAULT_SCAN_INTERVAL,
+    API_UPDATE_ERROR_MSG,
+    GOODWE_SPELLING,
+    AC_EMPTY,
+    AC_CURRENT_EMPTY,
+    AC_FEQ_EMPTY,
 )
-from homeassistant.helpers.entity import Entity
-from .const import DOMAIN, CONF_STATION_ID, DEFAULT_SCAN_INTERVAL
 
 _LOGGER = logging.getLogger(__name__)
 
 
-async def async_setup_entry(hass, config_entry, async_add_entities):
+def get_value_from_path(data, path):
+    value = data
+    for key in path:
+        value = value[key]
+    return value
+
+
+class Sensor(CoordinatorEntity, SensorEntity):
+    str_clean_regex = re.compile(r"(\d+\.?\d*)")
+
+    def __init__(
+        self,
+        coordinator,
+        device_info: DeviceInfo,
+        unique_id: str,
+        name: str,
+        value_path: List[str],
+        data_type_converter: Callable,
+        device_class: Optional[SensorDeviceClass] = None,
+        native_unit_of_measurement: Optional[str] = None,
+        state_class: Optional[SensorStateClass] = None,
+        empty_value=None,
+    ):
+        super().__init__(coordinator)
+        self._value_path = value_path
+        self._data_type_converter = data_type_converter
+        self._empty_value = empty_value
+
+        self._attr_unique_id = unique_id
+        self._attr_name = name
+        self._attr_native_unit_of_measurement = native_unit_of_measurement
+        self._attr_device_class = device_class
+        self._attr_state_class = state_class
+        self._attr_device_info = device_info
+
+    def _get_native_value_from_coordinator(self):
+        return get_value_from_path(self.coordinator.data, self._value_path)
+
+    @property
+    def native_value(self):
+        """Return the state of the device."""
+        value = get_value_from_path(self.coordinator.data, self._value_path)
+        if isinstance(value, str):
+            value = self.str_clean_regex.search(value).group(1)
+        if self._empty_value is not None and self._empty_value == value:
+            value = 0
```

Comment:
I receive `None` values for some attributes, e.g. `['inverter', 0, 'invert_full', 'vpv3']`. They cause an unhandled exception, so all updates of values failed. My quick fix is to detect `None` and treat it as `0`. You may instead consider a try-catch around the conversion.

```suggestion
        if self._empty_value is not None and self._empty_value == value:
            value = 0
        if None == value:
            _LOGGER.debug(f"Value was 'None': {self._value_path}")
            value = 0
```